File: orchestra/state_machine.py
"""Pure derivation of a task agent's lifecycle state."""

from __future__ import annotations

CLAIMED = "CLAIMED"
RUNNING = "RUNNING"
STOPPING = "STOPPING"
IDLE = "IDLE"
WAITING = "WAITING"
FAILED = "FAILED"
STOPPED = "STOPPED"
GAVE_UP = "GAVE_UP"


def derive_state(
    *,
    alive: bool,
    stopping: bool,
    gave_up: bool,
    turn: dict,
    status: dict | None,
    waits: bool,
    starting: bool = False,
) -> str:
    if alive:
        return STOPPING if stopping else RUNNING
    if gave_up:
        return GAVE_UP
    if stopping:
        return STOPPED
    if not turn:
        return CLAIMED
    if status is None or status.get("sequence") != turn.get("sequence"):
        return RUNNING if starting else FAILED
    if status.get("exit_code") != 0 or status.get("is_error"):
        return FAILED
    return WAITING if waits else IDLE

File: patchwake/state-machine.ts
import type { CompletionRecord, TaskState, TurnRecord } from "./models.js";

export const CLAIMED = "CLAIMED",
  RUNNING = "RUNNING",
  STOPPING = "STOPPING",
  IDLE = "IDLE";
export const WAITING = "WAITING",
  FAILED = "FAILED",
  STOPPED = "STOPPED",
  GAVE_UP = "GAVE_UP";

export function deriveState(value: {
  alive: boolean;
  stopping: boolean;
  gaveUp: boolean;
  turn: TurnRecord;
  status: CompletionRecord | null;
  waits: boolean;
  starting?: boolean;
}): TaskState {
  const {
    alive,
    stopping,
    gaveUp,
    turn,
    status,
    waits,
    starting = false,
  } = value;
  if (alive) return stopping ? STOPPING : RUNNING;
  if (gaveUp) return GAVE_UP;
  if (stopping) return STOPPED;
  if (Object.keys(turn).length === 0) return CLAIMED;
  if (status === null || status.sequence !== turn.sequence)
    return starting ? RUNNING : FAILED;
  if (status.exit_code !== 0 || status.is_error) return FAILED;
  return waits ? WAITING : IDLE;
}
